backends/cloud/basestorage.cpp

```diff
@@ -0,0 +1,214 @@
+			debug(9, "%s", json->stringify(true).c_str());
+			success = false;
+		}
+	}
+
+	if (success && result.getVal("error")->asBool()) {
+		Common::String errorMessage = "{error: true}, message is missing";
+		if (Networking::CurlJsonRequest::jsonContainsString(result, "message", "BaseStorage::codeFlowComplete")) {
+			errorMessage = result.getVal("message")->asString();
+		}
+		warning("BaseStorage: response says error occurred: %s", errorMessage.c_str());
+		success = false;
+		callbackMessage = errorMessage;
+	}
+
+	if (success && !Networking::CurlJsonRequest::jsonContainsObject(result, "oauth", "BaseStorage::codeFlowComplete")) {
+		warning("BaseStorage: bad response, no 'oauth' attribute passed");
+		debug(9, "%s", json->stringify(true).c_str());
+		success = false;
+	}
+
+	Common::JSONObject oauth;
+	bool requiresRefreshToken = needsRefreshToken();
+	if (success) {
+		oauth = result.getVal("oauth")->asObject();
+		if (!Networking::CurlJsonRequest::jsonContainsString(oauth, "access_token", "BaseStorage::codeFlowComplete") ||
+			!Networking::CurlJsonRequest::jsonContainsString(oauth, "refresh_token", "BaseStorage::codeFlowComplete", !requiresRefreshToken)) {
+			warning("BaseStorage: bad response, no 'access_token' or 'refresh_token' attribute passed");
+			debug(9, "%s", json->stringify(true).c_str());
+			success = false;
+		}
+	}
+
+	if (success) {
+		debug(9, "%s", json->stringify(true).c_str()); // TODO: remove when done testing against cloud.scummvm.org
+		_token = oauth.getVal("access_token")->asString();
+		if (requiresRefreshToken) {
+			_refreshToken = oauth.getVal("refresh_token")->asString();
+		}
+		CloudMan.replaceStorage(this, storageIndex());
+		ConfMan.flushToDisk();
+	}
+
+	if (!success)
+		CloudMan.removeStorage(this);
+	if (callback)
+		(*callback)(Networking::ErrorResponse(nullptr, false, !success, callbackMessage, -1));
+	delete json;
+	delete callback;
+}
+
+void BaseStorage::codeFlowFailed(Networking::ErrorCallback callback, Networking::ErrorResponse error) {
+	debug(9, "BaseStorage: code flow failed (%s, %ld):", (error.failed ? "failed" : "interrupted"), error.httpResponseCode);
+	debug(9, "%s", error.response.c_str());
+	CloudMan.removeStorage(this);
+
+	if (callback)
+		(*callback)(error);
+	delete callback;
+}
+
+void BaseStorage::refreshAccessToken(BoolCallback callback, Networking::ErrorCallback errorCallback) {
+	if (_refreshToken == "") {
+		warning("BaseStorage: no refresh token available to get new access token.");
+		if (callback) (*callback)(BoolResponse(nullptr, false));
+		return;
+	}
+
+	Networking::JsonCallback innerCallback = new Common::CallbackBridge<BaseStorage, BoolResponse, Networking::JsonResponse>(this, &BaseStorage::tokenRefreshed, callback);
+	if (errorCallback == nullptr)
+		errorCallback = getErrorPrintingCallback();
+
+	Common::String url = Common::String::format("https://cloud.scummvm.org/%s/refresh?code=%s", cloudProvider().c_str(), _refreshToken.c_str());
```

Comment:
I suggest passing the refresh token as a POST parameter (or a request header) so it is not stored in the web server logs.

Reply:
POST wouldn't be appropriate here, but if we want to move it to a header that's fine.

Reply:
I'd be glad to be enlightened as to why POST is not appropriate :)

Reply:
@Tkachov I'm going to move it from a query param to a header named `X-ScummVM-Refresh-Token`

Reply:
OK. I'm in for POST too though (not semantically appropriate for RESTful, but who really cares?)